server/src/gnome/prompter.rs

```diff
@@ -0,0 +1,450 @@
+    // SecretExchange aes_key
+    secret_exchange_aes_key: Arc<RwLock<String>>,
+    service: Service,
+    prompt_path: OwnedObjectPath,
+    path: OwnedObjectPath,
+}
+
+#[interface(name = "org.gnome.keyring.internal.Prompter.Callback")]
+impl PrompterCallback {
+    pub async fn prompt_ready(
+        &self,
+        reply: Optional<Reply>,
+        _properties: Properties,
+        exchange: &str,
+        #[zbus(connection)] connection: &zbus::Connection,
+    ) -> Result<(), ServiceError> {
+        let Some(prompt) = self.service.prompt(&self.prompt_path).await else {
+            return Err(ServiceError::NoSuchObject(
+                "Prompt does not exist.".to_string(),
+            ));
+        };
+
+        let prompt_role = prompt.role();
+        match prompt_role {
+            PromptRole::Lock => {
+                self.prompt(
+                    prompt_role,
+                    reply,
+                    exchange,
+                    connection.clone(),
+                    prompt.path().clone(),
+                    prompt.objects().clone(),
+                )
+                .await?;
+            }
+            PromptRole::Unlock => {
+                self.prompt(
+                    prompt_role,
+                    reply,
+                    exchange,
+                    connection.clone(),
+                    prompt.path().clone(),
+                    prompt.objects().clone(),
+                )
+                .await?;
+            }
+            PromptRole::CreateCollection => todo!(),
+        };
+
+        Ok(())
+    }
+
+    pub async fn prompt_done(
+        &self,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
+    ) -> Result<(), ServiceError> {
+        if let Some(prompt) = self.service.prompt(&self.prompt_path).await {
+            let path = prompt.path();
+            object_server.remove::<Prompt, _>(path).await?;
+            self.service.remove_prompt(path).await;
+        }
+        object_server.remove::<Self, _>(&self.path).await?;
+
+        Ok(())
+    }
+}
+
+impl PrompterCallback {
+    pub async fn new(window_id: &str, service: Service, prompt_path: OwnedObjectPath) -> Self {
+        let index = service.prompt_index().await;
+        Self {
+            window_id: window_id.to_owned(),
+            secret_exchange_pub_key: Default::default(),
+            secret_exchange_aes_key: Default::default(),
+            path: OwnedObjectPath::try_from(format!("/org/gnome/keyring/Prompt/p{index}")).unwrap(),
+            service,
+            prompt_path,
+        }
+    }
+
+    pub fn path(&self) -> &OwnedObjectPath {
+        &self.path
+    }
+
+    pub async fn object_label(&self, objects: &Vec<OwnedObjectPath>) -> Option<String> {
+        for object in objects {
+            if let Some(collection) = self.service.collection_from_path(object).await {
+                return Some(collection.label().await);
+            }
+        }
+        None
```

Comment:
The objects could be items, not necessarily collections, no?

Reply:
Yes. This is fixed now, thanks!